### ISSUE TITLE
Calculate orbitals for large system, get DFT energy for smaller system

### ISSUE THREAD

**Issue by aizvorski** [pyscf/pyscf#1973]
Hello - I have a scenario where I'd like to calculate orbitals and energy in a large system (let's say several molecules), then delete some molecules (both the nuclei and oribtals) and recalculate the energy of the rest (let's say a single molecule).  Essentially I want to reuse orbitals from calculation to calculation, but also subset them. Both calculations would use DFT.

Is it possible to do this in PySCF?

**Comment by tberkel** [comment]
Things like this are definitely possible with PySCF but you need to be a bit more specific in your goal. Typically, molecular orbitals are delocalized over the entire system. When you delete some atoms, what do you want to happen? One option would be to localize the orbitals before deleting some atoms. But then after deleting some atoms, do you want the orbitals on the remaining atoms to still be able to use the basis functions from the deleted atoms (this would be like using ghost atoms)? If there is a specific reference describing the calculation you want to do, we can probably help more quickly.

**Comment by aizvorski** [comment]
Hi @tberkel - Thanks for replying! Sure, I'll try to describe this. My calculation would have several molecules - I would want to delete whole molecules.  Yes, the orbitals are delocalized between molecules too :) but I was hoping to try to ignore that effect as a first approximation. So, I'd like to simply delete both atoms and orbitals entirely. Is that doable?

(If ignoring this doesn't produce good results, the other option is to try to project the initial set of orbitals onto the final set of orbitals - maybe use an orthogonal projection)

**Comment by aizvorski** [comment]
A minimal example would be a water dimer, like this xyz file

```
6

O         -1.62893       -0.04138        0.37137
H         -0.69803       -0.09168        0.09337
H         -2.06663       -0.73498       -0.13663
O          1.21457        0.03172       -0.27623
H          1.44927        0.91672       -0.58573
H          1.72977       -0.08038        0.53387
```

I'd like to calculate energy and orbitals with DFT for the dimer, then delete one water (first 3 atoms) and calculate the energy of the other water using the same orbitals as it had in the dimer, also with DFT.

**Comment by verena-neufeld** [comment]
I haven’t tested this but as long as the orbital energies get recalculated and therefore the ordering for occupancy gets redone, it might be ok

**Comment by tberkel** [comment]
Sorry, but I'm still not quite sure precisely what you'd like to happen, so I don't want to give you wrong advice. Molecular orbitals are linear combinations of all basis functions,

$$ 
\phi_i(r) = \sum_{\mu} C_{\mu i} \chi_i(r) 
$$

and each basis function belongs to an atom. When you delete some atoms, what do you want to happen to the coefficients associated with basis functions on those atoms? Set them to zero? (Note that I would call this removing basis functions, and not deleting orbitals). If yes, then you'll have to think about restoring orthonormalization of the molecular orbitals. And a second, related question: I suppose you want to keep the molecular orbitals that are "primarily" on the molecules you keep and get rid of the molecular orbitals that are not, but how do you want to make this decision? In general, molecular orbitals can be delocalized and it's not trivial to assign them to one molecule.

For what it's worth, I would suggest trying to find a method in the literature that's close to what you'd like to do. My guess is that it will involve orbital localization, and possibly some kind of embedding. Again it might help if you tell us your final goal. Why would you want to calculate the energy of a smaller system after calculating the energy of a larger system?

**Comment by aizvorski** [comment]
@tberkel Ah, I see what you mean.  I want to try the simplest version of this first: delete the atoms and their basis functions (or equivalently, set their coefficients to zero), and just keep the remaining basis function coefficients as is, without trying to restore orthonormalization.

This is probably pretty physically reasonable for the systems I'm thinking of - the orbitals on different molecules do interact, but they don't overlap very much.  So they could (eg) mutually polarize, and I want to keep that effect frozen; but they're not really delocalized across molecules at all.

**Comment by tberkel** [comment]
And maybe you're already familiar, but you could also check out existing energy decomposition analysis methods, e.g., ALMO-EDA sounds similar to what you're after and indeed requires localized molecular orbitals.

**Comment by susilehtola** [comment]
I would steer away from orbitals altogether. The SCF energy is fully defined by the density matrix, which is also invariant to occupied orbital transformations. You can just extract the blocks of the density matrix that survive on the atoms that have not been deleted, and evaluate the corresponding energy.

**Comment by januseriksen** [comment]
I agree with @susilehtola. In mean-field theory, you can choose to limit all trace operations in your energy expression to only those AOs that are spatially assigned to a given atom, akin to what Nakai has previously proposed:

https://www.sciencedirect.com/science/article/pii/S000926140201151X

Please note, however, that this sort of partitioning scheme will be very sensitive to your choice of AO basis. Things are like better done by means of spatially localized MOs instead, as @tberkel also mentioned above. See, for instance, our recent work:

https://pubs.acs.org/doi/10.1021/acs.jctc.2c01290
https://arxiv.org/abs/2308.07780

**Comment by aizvorski** [comment]
@susilehtola @januseriksen @tberkel  Thank you all for the advice!  I got something rough working, see code snippet below.  As you mentioned, it is very sensitive to the choice of basis set.  Probably the most obvious symptom is that about 0.1 electrons are lost by deleting blocks of the density matrix like this, and 0.1 electrons contribute a pretty large amount of energy.  That's much less of an issue at longer distances, but in a typical hydrogen bond it is a large effect.

How would I modify this to do use spatially localized MOs?  (eg the IBO/IAO from @januseriksen 's paper)

```
from pyscf import gto, dft
import numpy as np

atoms = [
['O',        [ -1.62893,       -0.04138,        0.37137 ]],
['H',        [ -0.69803,       -0.09168,        0.09337 ]],
['H',        [ -2.06663,       -0.73498,       -0.13663 ]],
['O',        [  1.21457,        0.03172,       -0.27623 ]],
['H',        [  1.44927,        0.91672,       -0.58573 ]],
['H',        [  1.72977,       -0.08038,        0.53387 ]],
]

basis_name = 'def2-svp'
xc_name = 'r2scan'

mol = gto.M(atom=atoms, basis=basis_name, verbose=5, output='pyscf_log.txt')
mf = dft.RKS(mol)
mf.xc = xc_name
mf.kernel()

num_atoms = mf.mol._atm.shape[0]
num_orbitals = mf.mo_coeff.shape[0]
nuclear_charges = np.array(mol._atm[:,0])

dm = mf.make_rdm1(mf.mo_coeff)
num_electrons_dimer = np.trace(dm)
e_dimer = mf.energy_tot(dm)

dm = mf.make_rdm1(mf.mo_coeff)
mf.mol._atm[:, 0] = nuclear_charges
mf.mol._atm[:num_atoms//2, 0] = 0
dm[:num_orbitals//2, :] = 0
dm[:, :num_orbitals//2] = 0
e_monomer_first = mf.energy_tot(dm)

dm = mf.make_rdm1(mf.mo_coeff)
mf.mol._atm[:, 0] = nuclear_charges
mf.mol._atm[num_atoms//2:, 0] = 0
dm[num_orbitals//2:, :] = 0
dm[:, num_orbitals//2:] = 0
e_monomer_second = mf.energy_tot(dm)

print(e_dimer, e_monomer_first, e_monomer_second)
# -152.64719593212305 -76.2745173297667 -76.32035436497154

# for comparison, just calculate each monomer separately
mol = gto.M(atom=atoms[:num_atoms//2], basis=basis_name)
mf = dft.RKS(mol)
mf.xc = xc_name
mf.kernel()
e_monomer_first_separate = mf.energy_tot()

mol = gto.M(atom=atoms[num_atoms//2:], basis=basis_name)
mf = dft.RKS(mol)
mf.xc = xc_name
mf.kernel()
e_monomer_second_separate = mf.energy_tot()

print(e_monomer_first_separate, e_monomer_second_separate)
# -76.317336944196 -76.31740209505912
```

**Comment by januseriksen** [comment]
Apologies, but I've lost track of what exactly you're interested in doing?

You can opt for both AO- or MO-based partitionings of the total energy (and then accumulate atomic contributions to achieve whatever decomposition you desire) by using the following code (built upon PySCF):

https://github.com/januseriksen/decodense

Both types of decompositions are lossless (exact), but the MO-based variant is way more robust than that based on the locality of the underlying AO basis.

**Comment by aizvorski** [comment]
@januseriksen Thanks - I was trying to look at the nonbonded energy and separately the polarization energy of molecules in a dimer/larger cluster by deleting one molecule at a time from the density matrix, using the original atomic orbitals (just as an test, to get some idea of how well it works).  Probably very similar idea to decondense, although it looks like in your case you're trying to get decompose energy to atoms or bonds? (and of course you're using MO which would work better)

It seems like the decondense readme says to look in the examples/ directory but that is not in the github repo, could you add it when you get a chance?

**Comment by tberkel** [comment]
And just in case you aren't aware, PySCF can perform various orbital localization methods, as described here: https://pyscf.org/user/lo.html with examples in the `examples/lo` folder.

**Comment by januseriksen** [comment]
Hi @aizvorski,

Please have a look in the `tests` directory of the decodense repo, and feel free to reach out to me via email if you have further questions related to AO- and MO-based decompositions in general.

If all of your PySCF-related questions have been satisfactorily answered, please consider closing this issue.

**Comment by sunqm** [comment]
Feel free to reopen the issue to continue the discussion.
